frame/gilt/src/lib.rs

```diff
@@ -0,0 +1,578 @@
+					T::Currency::unreserve(&gilt.who, gilt.amount);
+					let amount = gilt_value - gilt.amount;
+					let deficit = T::Currency::deposit_creating(&gilt.who, amount);
+					T::Deficit::on_unbalanced(deficit);
+				} else {
+					if gilt_value < gilt.amount {
+						// We take anything reserved beyond the gilt's final value.
+						let rest = gilt.amount - gilt_value;
+						// `slash` might seem a little aggressive, but it's the only way to do it
+						// in case it's locked into the staking system.
+						let surplus = T::Currency::slash_reserved(&gilt.who, rest).0;
+						T::Surplus::on_unbalanced(surplus);
+					}
+					// Unreserve only its new value (less than the amount reserved). Everything
+					// should add up, but (defensive) in case it doesn't, unreserve takes lower
+					// priority over the funds.
+					T::Currency::unreserve(&gilt.who, gilt_value);
+				}
+
+				let e = Event::GiltThawed(index, gilt.who, gilt.amount, gilt_value);
+				Self::deposit_event(e);
+			});
+
+			Ok(().into())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Attempt to enlarge our gilt-set from bids in order to satisfy our desired target amount
+		/// of funds frozen into gilts.
+		pub fn pursue_target(max_bids: u32) -> Weight {
+			let totals = ActiveTotal::<T>::get();
+			if totals.proportion < totals.target {
+				let missing = totals.target.saturating_sub(totals.proportion);
+
+				let total_issuance = T::Currency::total_issuance();
+				let nongilt_issuance: u128 = total_issuance.saturating_sub(totals.frozen)
+					.saturated_into();
+				let effective_issuance = totals.proportion.left_from_one()
+					.saturating_reciprocal_mul(nongilt_issuance);
+				let intake: BalanceOf<T> = (missing * effective_issuance).saturated_into();
+
+				let (bids_taken, queues_hit) = Self::enlarge(intake, max_bids);
+				let first_from_each_queue = T::WeightInfo::pursue_target_per_queue(queues_hit);
+				let rest_from_each_queue = T::WeightInfo::pursue_target_per_item(bids_taken)
+						.saturating_sub(T::WeightInfo::pursue_target_per_item(queues_hit));
+				first_from_each_queue + rest_from_each_queue
+			} else {
+				T::WeightInfo::pursue_target_noop()
+			}
+		}
+
+		/// Freeze additional funds from queue of bids up to `amount`. Use at most `max_bids`
+		/// from the queue.
+		///
+		/// Return the number of bids taken and the number of distinct queues taken from.
+		pub fn enlarge(
+			amount: BalanceOf<T>,
+			max_bids: u32,
+		) -> (u32, u32) {
+			let total_issuance = T::Currency::total_issuance();
+			let mut remaining = amount;
+			let mut bids_taken = 0;
+			let mut queues_hit = 0;
+			let now = frame_system::Module::<T>::block_number();
+
+			ActiveTotal::<T>::mutate(|totals| {
+				QueueTotals::<T>::mutate(|qs| {
+					for duration in (1..=T::QueueCount::get()).rev() {
+						if qs[duration as usize - 1].0 == 0 {
+							continue
+						}
+						let queue_index = duration as usize - 1;
+						let expiry = now + T::Period::get() * duration.into();
```

Comment:
potential overflow from a configuration of large period or large current block number

Reply:
This would be a grave misconfiguration, which basically means the intergration coder's error.